common/sgx/verifier.c

```diff
@@ -0,0 +1,462 @@
+#include <openenclave/attestation/sgx/verifier.h>
+#include <openenclave/bits/safemath.h>
+#include <openenclave/internal/raise.h>
+#include <openenclave/internal/report.h>
+#include <openenclave/internal/sgx/plugin.h>
+
+#include "../common.h"
+#include "endorsements.h"
+#include "quote.h"
+#ifndef OE_BUILD_ENCLAVE
+#include "../../host/sgx/sgxquoteprovider.h"
+#endif
+
+static oe_result_t _on_register(
+    oe_attestation_role_t* context,
+    const void* config_data,
+    size_t config_data_size)
+{
+    OE_UNUSED(context);
+    OE_UNUSED(config_data);
+    OE_UNUSED(config_data_size);
+
+#ifdef OE_BUILD_ENCLAVE
+    return OE_OK;
+#else
+    // Intialize the quote provider if we want to verify a remote quote.
+    // Note that we don't have the OE_LINK_SGX_DCAP_QL guard here since we
+    // don't need the sgx libraries to verify the quote. All we need is the
+    // quote provider.
+    return oe_initialize_quote_provider();
+#endif
+}
+
+static oe_result_t _on_unregister(oe_attestation_role_t* context)
+{
+    OE_UNUSED(context);
+    return OE_OK;
+}
+
+static void _free_claim(oe_claim_t* claim)
+{
+    oe_free(claim->name);
+    oe_free(claim->value);
+}
+
+static oe_result_t _free_claims_list(
+    oe_verifier_t* context,
+    oe_claim_t* claims,
+    size_t claims_length)
+{
+    OE_UNUSED(context);
+    if (!claims)
+        return OE_OK;
+
+    for (size_t i = 0; i < claims_length; i++)
+        _free_claim(&claims[i]);
+    oe_free(claims);
+    return OE_OK;
+}
+
+static oe_result_t _get_input_time(
+    const oe_policy_t* policies,
+    size_t policies_size,
+    oe_datetime_t** time)
+{
+    if (!policies)
+    {
+        *time = NULL;
+        return OE_OK;
+    }
+
+    for (size_t i = 0; i < policies_size; i++)
+    {
+        if (policies[i].type == OE_POLICY_ENDORSEMENTS_TIME)
+        {
+            if (policies[i].policy_size != sizeof(**time))
+                return OE_INVALID_PARAMETER;
+
+            *time = (oe_datetime_t*)policies[i].policy;
+            break;
+        }
+    }
+
+    return OE_OK;
+}
+
+static oe_result_t _verify_local_report(
+    const uint8_t* evidence_buffer,
+    size_t evidence_buffer_size)
+{
+    // Do a normal report verification on the enclave side.
+    // Local report verification is unsupported for host side.
+#ifdef OE_BUILD_ENCLAVE
+    return oe_verify_report(evidence_buffer, evidence_buffer_size, NULL);
+#else
+    OE_UNUSED(evidence_buffer);
+    OE_UNUSED(evidence_buffer_size);
+    return OE_UNSUPPORTED;
+#endif
+}
+
+static oe_result_t _add_claim(
+    oe_claim_t* claim,
+    void* name,
+    size_t name_size,
+    void* value,
+    size_t value_size)
+{
+    if (*((uint8_t*)name + name_size - 1) != '\0')
+        return OE_CONSTRAINT_FAILED;
+
+    claim->name = (char*)oe_malloc(name_size);
+    if (claim->name == NULL)
+        return OE_OUT_OF_MEMORY;
+    memcpy(claim->name, name, name_size);
+
+    claim->value_size = value_size;
```

Comment:
>claim->value_size = value_size; [](start = 4, length = 31)

Don't assign to claim->value_size until the allocation succeeds.

Reply:
fixed.